core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableProject.java

```diff
@@ -86,4 +103,86 @@ public Result implement(EnumerableRelImplementor implementor, Prefer pref) {
     // EnumerableCalcRel is always better
     throw new UnsupportedOperationException();
   }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null) {
+      return null;
+    }
+
+    final Mappings.TargetMapping map =
+        RelOptUtil.permutationIgnoreCast(
+            getProjects(), getInput().getRowType());
+
+    for (RelFieldCollation rc : collation.getFieldCollations()) {
+      if (!isCollationOnTrivialExpr(map, rc)) {
+        return null;
+      }
+    }
+
+    final RelCollation newCollation = collation.apply(map);
+    return Pair.of(required, ImmutableList.of(required.replace(newCollation)));
+  }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> deriveTraits(
+      final RelTraitSet childTraits, final int childId) {
+    RelCollation collation = childTraits.getCollation();
+    if (collation == null) {
+      return null;
+    }
+
+    int maxField = Math.max(getInput().getRowType().getFieldCount(), getProjects().size());
+    Mappings.TargetMapping mapping = Mappings
+        .create(MappingType.FUNCTION, maxField, maxField);
+    for (Ord<RexNode> node : Ord.zip(getProjects())) {
+      if (node.e instanceof RexInputRef) {
+        mapping.set(((RexInputRef) node.e).getIndex(), node.i);
+      } else if (node.e.isA(SqlKind.CAST)) {
+        final RexNode operand = ((RexCall) node.e).getOperands().get(0);
+        if (operand instanceof RexInputRef) {
+          mapping.set(((RexInputRef) operand).getIndex(), node.i);
+        }
+      }
+    }
+
+    List<RelFieldCollation> collationFieldsToDerive = new ArrayList<>();
+    for (RelFieldCollation rc : collation.getFieldCollations()) {
+      if (isCollationOnTrivialExpr(mapping, rc)) {
+        collationFieldsToDerive.add(rc);
+      } else {
+        break;
+      }
+    }
+
+    if (collationFieldsToDerive.size() > 0) {
+      final RelCollation newCollation = RelCollations.of(collationFieldsToDerive).apply(mapping);
+      return Pair.of(childTraits.replace(newCollation), ImmutableList.of(childTraits));
+    } else {
+      return null;
+    }
+  }
+
+  // Determine mapping between project input and output fields. If sort
+  // relies on non-trivial expressions, we can't push.
+  private boolean isCollationOnTrivialExpr(
+      Mappings.TargetMapping map, RelFieldCollation fc) {
+    if (map.getTargetOpt(fc.getFieldIndex()) < 0) {
+      return false;
+    }
+    final RexNode node = getProjects().get(fc.getFieldIndex());
```

Comment:
Try this query:
```sql
select comm, deptno, slacker from
  (select * from sales.emp order by COMM, DEPTNO, SLACKER limit 10) t 
order by comm, slacker;
```

Reply:
Yes. Good point. 

Need to add 
```
    if (passdown) {
      node = getProjects().get(fc.getFieldIndex());
    } else {
      node = getProjects().get(map.getTarget(fc.getFieldIndex()));
    }
```

The old tests passes because I actually used the first few columns in test cases, thus derive traits wasn't affected.